### PR TITLE
Generalize the mode option in pre-commit

### PR DESCRIPTION
**Pull request by RotemAmit**
## Related Issues
fixes:https://jira-dc.paloaltonetworks.com/browse/CIAC-8764

## Description
Support the option of having multiple modes per hook (not just nightly mode).

### CODE REVIEW

File: demisto_sdk/commands/coverage_analyze/coverage_report.py [hunk 1]
```diff
@@ -134,8 +134,8 @@ def coverage_diff_report(self) -> bool:
             min_cov = self.file_min_coverage(file_name)
             if min_cov > cov_precents:
                 logger.error(
-                    f"[red]Unit-tests for '{file_name}' must reach a coverage of at least {min_cov}% "
-                    f"(currently: {cov_precents}%).[/red]"
+                    f"[red]Unit-tests for '{file_name}' must reach a coverage of at least {min_cov} percents "
+                    f"(currently: {cov_precents} percents).[/red]"
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/coverage_analyze/coverage_report.py:138]:
```suggestion
                    f"[red]Unit-tests for '{file_name}' must reach a coverage of at least {min_cov} percent "
                    f"(currently: {cov_precents} percent).[/red]"
```
but why this change?

File: demisto_sdk/commands/pre_commit/hooks/general_hook.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+from demisto_sdk.commands.pre_commit.hooks.hook import Hook
+
+
+class GeneralHook(Hook):
+    def prepare_hook(self, **kwargs):
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/hooks/general_hook.py:5]: we can make this the default in Hook and then just use hook. we dont really need this class

File: demisto_sdk/commands/pre_commit/hooks/hook.py [hunk 3]
```diff
@@ -24,6 +24,7 @@ def __init__(
         self.mode = mode
         self.all_files = all_files
         self.input_mode = input_mode
+        self._set_properties(hook={}, to_delete=to_delete)
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/hooks/hook.py:27]: to_delete cant be used properly in the init method, since it would delete all the properties before the hook has a change to process them. We should get rid of the deleting from this function, and offer a utility function to delete fields after the logic is done

Please update the docker class to fit with this standard

File: demisto_sdk/commands/pre_commit/hooks/validate_format.py [hunk 4]
```diff
@@ -13,14 +16,14 @@ def prepare_hook(self, files_to_run: Optional[Iterable[Path]], **kwargs):
         In case of an input or all files without nightly, runs validate/format on the given files.
         Otherwise runs validate/format with the -g flag.
         Args:
-            input_files (Optional[Iterable[Path]]): The input files to validate. Defaults to None.
+            files_to_run (Optional[Iterable[Path]]): The input files to validate. Defaults to None.
         """
-        if self.mode == PreCommitModes.NIGHTLY and self.all_files:
-            self.base_hook["args"].append("-a")
+        if self.mode and self.mode.lower() == "nightly" and self.all_files:
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/hooks/validate_format.py:21]: this couldnt be fixed with modes?

File: demisto_sdk/commands/pre_commit/pre_commit_command.py [hunk 5]
```diff
@@ -151,23 +151,34 @@ def prepare_hooks(
...
+        ValidateFormatHook(**hooks.pop("format"), **kwargs).prepare_hook(
             self.files_to_run
         )
-        ValidateFormatHook(**hooks["format"], **kwargs).prepare_hook(self.files_to_run)
         [
             DockerHook(**hook, **kwargs).prepare_hook(files_to_run=self.files_to_run)
             for hook_id, hook in hooks.items()
             if hook_id.endswith("in-docker")
         ]
+        docker_without_hooks = [
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/pre_commit_command.py:175]: this will include format and verify hooks and all the others. why only specifically exclude docker?

**Comment by JudahSchwartz** [comment]
I didnt review the coverage code itself

File: demisto_sdk/commands/coverage_analyze/coverage_report.py (hunk 1 above)
Comment (RotemAmit) [reply demisto_sdk/commands/coverage_analyze/coverage_report.py:138]: @ilaner and I made this change because there was a bug since we can't use % in the logger due to the format string.

File: demisto_sdk/commands/pre_commit/hooks/general_hook.py (hunk 2 above)
Comment (RotemAmit) [reply demisto_sdk/commands/pre_commit/hooks/general_hook.py:5]: I tried using Hook, but I got an error that it is not possible to instantiate an abstract class, and I must use a class that inherits Hook.

File: demisto_sdk/commands/pre_commit/pre_commit_command.py (hunk 5 above)
Comment (RotemAmit) [reply demisto_sdk/commands/pre_commit/pre_commit_command.py:175]: in all of the other hooks (PyclnHook, RuffHook, MypyHook, SourceryHook, ValidateFormatHook) I used `pop()` as we discussed, and I also need to exclude the docker hooks as well before initializing the remaining hooks.

File: demisto_sdk/commands/pre_commit/tests/pre_commit_test.py [hunk 6]
```diff
@@ -369,6 +371,44 @@ def test_exclude_python2_of_non_supported_hooks(mocker, repo: Repo):
...
+        "-i",
+        ".coverage",
+        "--report-dir",
+        "coverage_report",
+        "--report-type",
+        "all",
+        "--allowed-coverage-degradation-percentage",
+        "100",
+    ]
+    coverage_analyze_hook = create_hook({"args": args, "args:nightly": args_nightly})
```
Comment (eyalpalo) [inline demisto_sdk/commands/pre_commit/tests/pre_commit_test.py:404]: Lets move those args/args_nightly to be part of the parametrize.

File: demisto_sdk/commands/pre_commit/hooks/docker.py [hunk 7]
```diff
@@ -281,7 +282,10 @@ def get_new_hooks(
             if self._set_files_on_hook(hook, files):
                 ret_hooks.append(hook)
         for hook in ret_hooks:
-            self._set_properties(hook, to_delete=["docker_image", "config_file_arg"])
+            if hook.get("docker_image"):
+                hook.pop("docker_image")
+            if hook.get("config_file_arg"):
+                hook.pop("config_file_arg")
```
Comment (eyalpalo) [inline demisto_sdk/commands/pre_commit/hooks/docker.py:288]: Do we have a test that tests the final formation of the hooks dict?

File: demisto_sdk/commands/pre_commit/README.md [hunk 8]
```diff
@@ -18,6 +18,21 @@ It utilizes the [pre-commit](https://github.com/pre-commit/pre-commit) infrastru
 ### Automatically, as a git hook
 * Create a [git hook](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) that calls `demisto-sdk pre-commit`.
 
+## Modes
+If you need different args set for the different modes, for example some rules should be excluded in the nightly build.
```
Comment (eyalpalo) [inline demisto_sdk/commands/pre_commit/README.md:22]:
```suggestion
When different args set for the different modes are needed, for example, some rules should be excluded in the nightly build.
```

**Review by eyalpalo — CHANGES_REQUESTED** [review]

**Review by eyalpalo — APPROVED** [review]

File: demisto_sdk/commands/coverage_analyze/coverage_report.py (hunk 1 above)
Comment (JudahSchwartz) [reply demisto_sdk/commands/coverage_analyze/coverage_report.py:138]: still should be percent not percents

File: demisto_sdk/commands/pre_commit/hooks/hook.py [hunk 9]
```diff
@@ -24,15 +22,15 @@ def __init__(
         self.mode = mode
         self.all_files = all_files
         self.input_mode = input_mode
+        self._set_properties(hook={})
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/hooks/hook.py:25]: i think we can get rid of the hook arg

File: demisto_sdk/commands/pre_commit/pre_commit_command.py [hunk 10]
```diff
@@ -151,23 +150,32 @@ def prepare_hooks(
             "all_files": self.all_files,
             "input_mode": self.input_mode,
         }
-        PyclnHook(**hooks["pycln"], **kwargs).prepare_hook(PYTHONPATH)
-        RuffHook(**hooks["ruff"], **kwargs).prepare_hook(
+        PyclnHook(**hooks.pop("pycln"), **kwargs).prepare_hook(PYTHONPATH)
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/pre_commit_command.py:153]: these will still fail if the hooks arent present. We should enable not having these hooks in the config. Pop with None as default return

File: demisto_sdk/commands/coverage_analyze/coverage_report.py (hunk 1 above)
Comment (RotemAmit) [reply demisto_sdk/commands/coverage_analyze/coverage_report.py:138]: fixed

File: demisto_sdk/commands/pre_commit/pre_commit_command.py [hunk 11]
```diff
@@ -152,25 +151,34 @@ def prepare_hooks(
             "all_files": self.all_files,
             "input_mode": self.input_mode,
         }
-        PyclnHook(**hooks["pycln"], **kwargs).prepare_hook(PYTHONPATH)
-        RuffHook(**hooks["ruff"], **kwargs).prepare_hook(
+        PyclnHook(**hooks.pop("pycln", None), **kwargs).prepare_hook(PYTHONPATH)
+        RuffHook(**hooks.pop("ruff", None), **kwargs).prepare_hook(
             self.python_version_to_files, IS_GITHUB_ACTIONS
         )
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/pre_commit/pre_commit_command.py:157]: will passing None to these hooks work as expected? Shouldnt we check if we got these hooks in the yml?

**Review by JudahSchwartz — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/pre_commit/pre_commit_command.py (hunk 11 above)
Comment (JudahSchwartz) [reply demisto_sdk/commands/pre_commit/pre_commit_command.py:157]: meaning 
```
hook = hooks.pop(ruff,None)
if hook: 
    ruffHook(.....
```

**Review by JudahSchwartz — APPROVED** [review]
great work on this 1. nice end product despite the changing requirements 🚀
